test/e2e_node/device_manager_test.go

```diff
@@ -266,6 +273,211 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+		variable: AUTO_REGISTER. By default AUTO_REGISTER=true allowing the device plugin
+		to autoregister to kubelet. For this e2e test, we use sample device plugin
+		spec with AUTO_REGISTER=false to first manually register the plugin to allow
+		an application pod (requesting devices) to successfully run on the node
+		followed by kubelet restart where device plugin doesn't register and the
+		application pod fails with admission error.
+
+		   Breakdown of the steps implemented as part of this e2e test is as follows:
+		   1. Create sample device plugin with AUTO_REGISTER= false which creates a
+		      directory at `/var/lib/kubelet/device-plugins/registered` and waits
+			  for a client to delete that directory.
+		   2. Trigger plugin registeration by deleting the abovementioned directory.
+		   3. Create a test pod requesting devices exposed by the device plugin.
+		   4. Stop kubelet.
+		   5. Remove pods using CRI to ensure new pods are created after kubelet restart.
+		   6. Restart kubelet.
+		   7. Wait for the sample device plugin pod to be running. In this case,
+		      the registration is not triggered.
+		   8. Ensure that resource capacity/allocatable exported by the device plugin is zero.
+		   9. The test pod should fail with `UnexpectedAdmissionError`
+		   10. Delete the test pod.
+		   11. Delete the sample device plugin pod.
+	*/
+	ginkgo.Context("With sample device plugin", func() {
+		var deviceCount int = 2
+		var devicePluginPod *v1.Pod
+
+		// this test wants to reproduce what happened in https://github.com/kubernetes/kubernetes/issues/109595
+		ginkgo.BeforeEach(func(ctx context.Context) {
+			ginkgo.By("Wait for node to be ready")
+			gomega.Eventually(ctx, e2enode.TotalReady).
+				WithArguments(ctx, f.ClientSet).
+				WithTimeout(time.Minute).
+				Should(gomega.BeEquivalentTo(1))
+
+			ginkgo.By("Scheduling a sample device plugin pod")
+			data, err := e2etestfiles.Read(SampleDevicePluginControlRegistrationDSYAML)
+			if err != nil {
+				framework.Fail(err.Error())
+			}
+			ds := readDaemonSetV1OrDie(data)
+
+			dp := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: SampleDevicePluginName,
+				},
+				Spec: ds.Spec.Template.Spec,
+			}
+
+			devicePluginPod = e2epod.NewPodClient(f).CreateSync(ctx, dp)
+
+			go func() {
+				// Since autoregistration is disabled for the device plugin (as AUTO_REGISTER=false),
+				// device plugin registration needs to be triggerred manually.
+				// This is done by deleting directory created by sample device plugin:
+				// `/var/lib/kubelet/device-plugins/registered`.
+
+				defer ginkgo.GinkgoRecover()
+				triggerPath := devicePluginDir + "/registered"
+				err := os.RemoveAll(triggerPath)
+				framework.ExpectNoError(err)
+			}()
+
+			ginkgo.By("Waiting for devices to become available on the local node")
+
+			gomega.Eventually(ctx, isNodeReadyWithSampleResources).
+				WithArguments(ctx, f).
```

Comment:
Don't pass `ctx` as argument. `gomega.Eventually` will handle that automatically.

I find it easier to write:
```
gomega.Eventually(ctx, isNodeReadyWithSampleResources, f).
```
without the `WithArguments`

Reply:
This also applies below.

Reply:
Thanks, I am not passing `ctx` as argument in the `WithArguments` anymore. 

I tried playing around with the approach of not using `WithArguments` as per your suggestion but the test fails without it.

Reply:
Fails how? I'm curious because the approach with directly passing the arguments is used everywhere else, `WithArguments` is not used anywhere.

Reply:
Yep, I thought so too when I referred to other areas in the code base that used eventually the way you suggested. Here is what I get:

```
  Feb 28 19:06:07.615: INFO: 
  Logging kubelet events for node swsehgal-node
  Feb 28 19:06:07.618: INFO: 
  Logging pods the kubelet thinks is on node swsehgal-node
  W0228 19:06:07.623862  111282 metrics_grabber.go:111] Can't find any pods in namespace kube-system to grab metrics from
  Feb 28 19:06:07.657: INFO: 
  Latency metrics for node swsehgal-node
  STEP: Destroying namespace "devicemanager-test-3328" for this suite. @ 02/28/23 19:06:07.658
  << Timeline

  [FAILED] &framework.Framework{BaseName:"devicemanager-test", UniqueName:"devicemanager-test-3328", clientConfig:(*rest.Config)(0xc001ab0240), ClientSet:(*kubernetes.Clientset)(0xc0001a0b60), KubemarkExternalClusterClientSet:kubernetes.Interface(nil), DynamicClient:(*dynamic.DynamicClient)(0xc00061d570), ScalesGetter:(*scale.scaleClient)(0xc001b09a80), SkipNamespaceCreation:false, Namespace:(*v1.Namespace)(0xc000f68420), namespacesToDelete:[]*v1.Namespace{(*v1.Namespace)(0xc000f68420)}, NamespaceDeletionTimeout:0, NamespacePodSecurityEnforceLevel:"privileged", flakeReport:(*framework.FlakeReport)(0xc001b09dc0), Options:framework.Options{ClientQPS:20, ClientBurst:50, GroupVersion:(*schema.GroupVersion)(nil)}, TestSummaries:[]framework.TestDataSummary(nil), Timeouts:(*framework.TimeoutContext)(0xc000ecb180), DumpAllNamespaceInfo:(framework.DumpAllNamespaceInfoAction)(0x4679320)} is not a valid interval. Must be a time.Duration, a parsable duration string, or a number.
  In [BeforeEach] at: test/e2e_node/device_manager_test.go:386 @ 02/28/23 19:06:03.575
------------------------------
SSSSSSSSSSSSS

Summarizing 1 Failure:
  [FAIL] [sig-node] Device Manager  [Serial] [Feature:DeviceManagerBug][NodeFeature:DeviceManager] With sample device plugin [BeforeEach] should deploy pod consuming devices first but fail with admission error after kubelet restart in case device plugin hasn't re-registered
  test/e2e_node/device_manager_test.go:386

Ran 1 of 389 Specs in 16.926 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 388 Skipped
--- FAIL: TestE2eNode (16.96s)
```

Reply:
The only change I made was:
```
diff --git a/test/e2e_node/device_manager_test.go b/test/e2e_node/device_manager_test.go
index 68eeb85b2d2..6fc60fc5ed0 100644
--- a/test/e2e_node/device_manager_test.go
+++ b/test/e2e_node/device_manager_test.go
@@ -383,8 +383,7 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
 
                        ginkgo.By("Waiting for devices to become available on the local node")
 
-                       gomega.Eventually(ctx, isNodeReadyWithSampleResources).
-                               WithArguments(f).
+                       gomega.Eventually(ctx, isNodeReadyWithSampleResources, f).
                                WithTimeout(5 * time.Minute).
                                Should(BeReady())
```

Reply:
I was wrong about  `Eventually` taking a function + parameters. The additional arguments after the function are timeout and poll interval. Your approach with `WithArguments` is correct.

Reply:
Thanks for double checking this.